src/Elastic.Apm/Report/PayloadSenderV2.cs

```diff
@@ -97,56 +104,85 @@ internal class PayloadSenderV2 : IPayloadSender, IDisposable
 
 			// Replace invalid characters by underscore. All invalid characters can be found at
 			// https://github.com/dotnet/corefx/blob/e64cac6dcacf996f98f0b3f75fb7ad0c12f588f7/src/System.Net.Http/src/System/Net/Http/HttpRuleParser.cs#L41
-			string AdaptUserAgentValue(string value) => Regex.Replace(value, "[ /()<>@,:;={}?\\[\\]\"\\\\]", "_");
+			string AdaptUserAgentValue(string value)
+			{
+				return Regex.Replace(value, "[ /()<>@,:;={}?\\[\\]\"\\\\]", "_");
+			}
 		}
 
 		private CancellationTokenSource _batchBlockReceiveAsyncCts;
 
-		public void QueueTransaction(ITransaction transaction)
-		{
-			var res = _eventQueue.Post(transaction);
-			_logger.Debug()
-				?.Log(!res
-					? "Failed adding Transaction to the queue, {Transaction}"
-					: "Transaction added to the queue, {Transaction}", transaction);
+		public void QueueTransaction(ITransaction transaction) => QueueEvent(transaction, "Transaction");
 
-			_eventQueue.TriggerBatch();
-		}
+		public void QueueSpan(ISpan span) => QueueEvent(span, "Span");
 
-		public void QueueSpan(ISpan span)
-		{
-			var res = _eventQueue.Post(span);
-			_logger.Debug()
-				?.Log(!res
-					? "Failed adding Span to the queue, {Span}"
-					: "Span added to the queue, {Span}", span);
-		}
+		public void QueueMetrics(IMetricSet metricSet) => QueueEvent(metricSet, "MetricSet");
 
-		public void QueueMetrics(IMetricSet metricSet)
-		{
-			var res = _eventQueue.Post(metricSet);
-			_logger.Debug()
-				?.Log(!res
-					? "Failed adding MetricSet to the queue, {MetricSet}"
-					: "MetricSet added to the queue, {MetricSet}", metricSet);
-		}
+		public void QueueError(IError error) => QueueEvent(error, "Error");
 
-		public void QueueError(IError error)
+		public void QueueEvent(object eventObj, string dbgEventKind)
 		{
-			var res = _eventQueue.Post(error);
-			_logger.Debug()
-				?.Log(!res
-					? "Failed adding Error to the queue, {Error}"
-					: "Error added to the queue, {Error}", error);
+			var addedSuccessfully = _eventQueue.Post(eventObj);
+
+			if (addedSuccessfully)
+			{
+				_logger.Debug()?.Log(dbgEventKind + " added to the queue. " + dbgEventKind + ": {" + dbgEventKind + "}", eventObj);
+
+				if (_flushInterval == TimeSpan.Zero)
+					_eventQueue.TriggerBatch();
+			}
+			else
+				_logger.Debug()?.Log("Failed adding " + dbgEventKind + " to the queue. " + dbgEventKind + ": {" + dbgEventKind + "}", eventObj);
 		}
 
 		private async Task DoWork()
 		{
 			_batchBlockReceiveAsyncCts = new CancellationTokenSource();
```

Comment:
Thank you for drawing my attention to this - I've noticed that the new tests slow down considerably and now I see that it's because each test leaks another `ElasticApmPayloadSender` thread. I've fixed `PayloadSenderV2.Dispose` to cancel all the ongoing work and stop the thread.

Reply:
Nice!